src/outputs/ascent.cpp

```diff
@@ -0,0 +1,203 @@
+#include "outputs/outputs.hpp"
+#include "utils/error_checking.hpp"
+
+// Ascent headers
+#ifdef PARTHENON_ENABLE_ASCENT
+#include "ascent.hpp"
+#include "conduit_blueprint.hpp"
+#include "conduit_relay_io.hpp"
+#include "conduit_relay_io_blueprint.hpp"
+#endif // ifdef PARTHENON_ENABLE_ASCENT
+
+namespace parthenon {
+
+//----------------------------------------------------------------------------------------
+//! \fn void AscentOutput:::WriteOutputFile(Mesh *pm)
+//  \brief  Expose mesh and all Cell variables for processing with Ascent
+void AscentOutput::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm,
+                                   const SignalHandler::OutputSignal signal) {
+#ifndef PARTHENON_ENABLE_ASCENT
+  if (Globals::my_rank == 0) {
+    PARTHENON_WARN("Ascent output requested by input file, but Ascent support not "
+                   "compiled in. Skipping this output type.");
+  }
+#else
+  // reference:
+  // https://llnl-conduit.readthedocs.io/en/latest/blueprint_mesh.html#complete-uniform-example
+
+  using conduit::Node;
+
+  // Ascent needs the MPI communicator we are using
+  ascent::Ascent ascent;
+  Node ascent_opts;
+  ascent_opts["mpi_comm"] = MPI_Comm_c2f(MPI_COMM_WORLD);
+  ascent_opts["actions_file"] = pin->GetString(output_params.block_name, "actions_file");
+  // Only publish fields that are used within actions to reduce memory footprint.
+  // A user might need to override this, e.g., in a runtime ascent_options.yaml, if
+  // the required fields cannot be resolved by Ascent.
+  // See https://ascent.readthedocs.io/en/latest/AscentAPI.html#field-filtering
+  // ascent_opts["field_filtering"] = "true";
+  ascent.open(ascent_opts);
+
+  // create root node for the whole mesh
+  Node root;
+
+  for (auto &pmb : pm->block_list) {
+    // create a unique id for this MeshBlock
+    const std::string &meshblock_name = "domain_" + std::to_string(pmb->gid);
+    Node &mesh = root[meshblock_name];
+
+    // add basic state info
+    mesh["state/domain_id"] = pmb->gid;
+    mesh["state/cycle"] = tm->ncycle;
+    mesh["state/time"] = tm->time;
+
+    auto &bounds = pmb->cellbounds;
+    auto ib = bounds.GetBoundsI(IndexDomain::entire);
+    auto jb = bounds.GetBoundsJ(IndexDomain::entire);
+    auto kb = bounds.GetBoundsK(IndexDomain::entire);
+    int nx = ib.e - ib.s + 1;
+    int ny = jb.e - jb.s + 1;
+    int nz = kb.e - kb.s + 1;
+    uint64_t ncells = nx * ny * nz;
+
+    auto ib_int = bounds.GetBoundsI(IndexDomain::interior);
+    auto jb_int = bounds.GetBoundsJ(IndexDomain::interior);
+    auto kb_int = bounds.GetBoundsK(IndexDomain::interior);
+
+    auto &coords = pmb->coords;
+    Real dx1 = coords.CellWidth<X1DIR>(ib.s, jb.s, kb.s);
+    Real dx2 = coords.CellWidth<X2DIR>(ib.s, jb.s, kb.s);
+    Real dx3 = coords.CellWidth<X3DIR>(ib.s, jb.s, kb.s);
+    std::array<Real, 3> corner = coords.GetXmin();
+
+    // create the coordinate set
+    mesh["coordsets/coords/type"] = "uniform";
```

Comment:
Assume this will have to change for non-Cartesian coords?

Reply:
As above, if you agree I'd add a loud fail here, too.

Reply:
This one can can be a loud fail.

Reply:
done